registry/distribution/repository.go

```diff
@@ -0,0 +1,674 @@
+
+	if resp.StatusCode != http.StatusCreated {
+		return parseErrorResponse(resp)
+	}
+	return verifyContentDigest(resp, expected.Digest)
+}
+
+// parseReference validates the reference.
+// Both simplified or fully qualified references are accepted.
+func (r *Repository) parseReference(reference string) (registry.Reference, error) {
+	ref, err := registry.ParseReference(reference)
+	if err != nil {
+		ref = registry.Reference{
+			Registry:   r.Reference.Registry,
+			Repository: r.Reference.Repository,
+			Reference:  reference,
+		}
+		if err = ref.ValidateReference(); err != nil {
+			return registry.Reference{}, err
+		}
+		return ref, nil
+	}
+	if ref.Registry == r.Reference.Registry && ref.Repository == r.Reference.Repository {
+		return ref, nil
+	}
+	return registry.Reference{}, fmt.Errorf("%w %q: expect %q", errdef.ErrInvalidReference, ref, r.Reference)
+}
+
+// Tags lists the tags available in the repository.
+func (r *Repository) Tags(ctx context.Context, fn func(tags []string) error) error {
+	url := fmt.Sprintf("%s/tags/list", r.endpoint())
+	if r.TagListPageSize > 0 {
+		url = fmt.Sprintf("%s?n=%d", url, r.TagListPageSize)
+	}
+
+	var err error
+	for err == nil {
+		url, err = r.tags(ctx, fn, url)
+	}
+	if err != errNoLink {
+		return err
+	}
+	return nil
+}
+
+// tags returns a single page of tag list with the next link.
+func (r *Repository) tags(ctx context.Context, fn func(tags []string) error, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := r.Client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", parseErrorResponse(resp)
+	}
+	var page struct {
+		Tags []string `json:"tags"`
+	}
+	lr := limitReader(resp.Body, r.MaxMetadataBytes)
+	if err := json.NewDecoder(lr).Decode(&page); err != nil {
+		return "", err
+	}
+	if err := fn(page.Tags); err != nil {
+		return "", err
+	}
+
+	return parseLink(resp)
+}
+
+// UpEdges returns the manifest descriptors directly referencing the given
+// manifest descriptor.
+// Reference: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md
+func (r *Repository) UpEdges(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	var res []ocispec.Descriptor
+	if err := r.Referrers(ctx, desc, func(referrers []artifactspec.Descriptor) error {
+		for _, referrer := range referrers {
+			res = append(res, descriptor.ArtifactToOCI(referrer))
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Referrers returns the manifest descriptors directly referencing the given
+// manifest descriptor.
+// Reference: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md
+func (r *Repository) Referrers(ctx context.Context, desc ocispec.Descriptor, fn func(referrers []artifactspec.Descriptor) error) error {
```

Comment:
Once https://github.com/opencontainers/distribution-spec/pull/111  is supported we need to revisit this path to return an error that registry extension is not supported if the request is a 404 and cache the response if possible. 
You can ignore this comment until that part of the PR is checked in.